### ISSUE TITLE
[rllib] Random seed for network initialization

### ISSUE THREAD

**Issue by whikwon** [ray-project/ray#2776]
### System information
- **OS Platform and Distribution (e.g., Linux Ubuntu 16.04)**: Ubuntu 16.04
- **Ray installed from (source or binary)**: source
- **Ray version**: 0.5.0
- **Python version**: 3.6.1
- **Exact command to reproduce**:

### Describe the problem

Is there controllable random seed parameter in rllib?

**Comment by ericl** [comment]
@richardliaw I recall you had a patch for this we haven't merged?

**Comment by ericl** [comment]
I see, then the main piece would be calling seed based on some configuration. This would be nice to have for the synchronous algorithms at least.

**Comment by whikwon** [comment]
@richardliaw  @ericl  Thank you for the comment. I'll use that utility for a while before being merged.

**Comment by hagrid67** [comment]
The utility referred to above just sets the seed for Tensorflow, Numpy and python random.  It doesn't provide a mechanism for choosing a value which is both unique across worker processes, and deterministic.  In the MPI-based OpenAI Baselines, they tend to use the MPI "rank", which is just an integer counting from zero assigned to each process started by the mpirun command, so in each process you can just do this:
`set_global_seeds(seed + 10000 * rank)`

It's not obvious to me what we could use instead of MPI rank in Ray.  For example, worker PID is not deterministic, and not even guaranteed unique across a cluster.

A further refinement: I'm just trying out the **Tune tutorial**; having started up a cluster, it runs training sessions with different hyperparams, each with say 8 workers out of a total of 32 in the cluster.  In this case, we should be using unique seeds in each of the 8 workers (say 1000 * worker number 0..7), and then replicating those seeds in each of the concurrent sessions, ie:
- workers 0..7 - session 1 - seeds 0, 1000, ... 7000
- workers 8..15 - session 2 - seeds 0, 1000, ... 7000
etc
...so that each training session with different hyperparams uses the same set of seeds.

I would also like to repeat the training session with a given set of hyperparams,  with several different sets of seeds, some thing like this:
`seed = repeat_run_number * 10  +  worker_number_in_session * 1000`
...so we can see the distribution of training progress across different seeds, with a given set of hyperparams.  (Is there already support within Tune for repeating same-hyperparam runs with different seeds?)

You seem to imply in some of the doc that Ray / Rllib supports **system restarts** within a cluster.  It looks even more difficult to try to support deterministic pseudo-random sequences across restarts. 

Any ideas for something to get me started?  Thanks!

**Comment by ericl** [comment]
> It's not obvious to me what we could use instead of MPI rank in Ray. For example, worker PID is not deterministic, and not even guaranteed unique across a cluster.

 You can do this with the worker_index property when configuring the env: https://ray.readthedocs.io/en/latest/rllib-env.html#configuring-environments

>...so we can see the distribution of training progress across different seeds, with a given set of hyperparams. (Is there already support within Tune for repeating same-hyperparam runs with different seeds?)

You can use `tune.sample_from(function)` to draw from a value from a function several times for the same grid search value. The number of samples is the `num_samples` config in tune: https://ray.readthedocs.io/en/latest/tune-usage.html#sampling-multiple-times

> You seem to imply in some of the doc that Ray / Rllib supports system restarts within a cluster. It looks even more difficult to try to support deterministic pseudo-random sequences across restarts.

I think the key is to make sure the seeds are tied back to the hyperparameter config (i.e. seed is a hyperparameter). That way it is always reproducible on restart.

**Comment by ericl** [comment]
cc @samp830 would be great if you could take a look at this. Looks like the main things to seed are:
- the model initialization
- gym env

**Comment by vonHartz** [comment]
@ericl  In the mean time, is there a way to patch the seed generation manually?  Which would be the places to do this? I tried to seed ray to train an agent deterministically , but could not make it work, yet. I tried doing it right after ray init and also tried patching the tune.run() function. Both apparently did not work.

**Comment by bmazoure** [comment]
It seems like if you make a wrapper around a `Trainer` class, and add
```
 def _init(self, config, env_creator):
            super()._init(config, env_creator)
            np.random.seed(seed)
            random.seed(seed)
            tf.set_random_seed(seed)
            torch.manual_seed(seed)
            if cuda:
                torch.cuda.manual_seed(seed)
```
the actions are reproducible, as far as I tested. To fix the env seed, an `gym.ObservationWrapper` with the line `env.seed(seed)` in the `__init__` method seems to do the job.

**Comment by richardliaw** [comment]
@bmazoure This is really neat! How much have you tested for this?

**Comment by bmazoure** [comment]
I tested it on value-based methods in PyTorch. If you don't put this chunk of code, the torch seed is reset every time (I was printing out the epsilon-greedy choices at the beginning of every episode). It wasn't very thorough, but so far I don't seem to have any issues. Then again, might be because I only tried on deterministic environments?

**Comment by ericl** [comment]
Fixed in https://github.com/ray-project/ray/pull/5197
